### PR TITLE
mesh: add mark_entities() method and RelabeledMesh() function

### PR DESCRIPTION
**Pull request by ksagiyam**
Allow for marking entities using indicator functions. This will be useful when:

- creating a mesh with modified labels
- making submeshes in the future

~TODO: make high-level interface: `RelabeledMesh()`.~ Done.

### CODE REVIEW

File: firedrake/mesh.py [hunk 1]
```diff
@@ -2601,6 +2675,95 @@ def _on_rank_vertices(vertex_coords, parent_mesh):
...
+    :kwarg cell_indicators: list of indicator functions (DP0 or DQ0) that mark
+        selected cells; each function marks interested cells as 1.
+    :kwarg cell_indicator_labels: list of integers that enumerates cell_indicators;
+        cell_indicators and cell_indicator_labels must have the same length.
+    :kwarg facet_indicators: list of indicator functions (HDiv Trace 0) that mark
+        selected facets; each function marks interested facets as 1.
+    :kwarg facet_indicator_labels: list of integers that enumerates facet_indicators;
+        facet_indicators and facet_indicator_labels must have the same length.
+    :kwarg name: optional name of the output mesh object.
+    """
```
Comment (ksagiyam) [inline firedrake/mesh.py:2692]: Do not need separate lists for cell and facet indicator functions.

Comment (ksagiyam) [reply firedrake/mesh.py:2690]: Fixed.

File: firedrake/cython/dmcommon.pyx [hunk 2]
```diff
@@ -3166,3 +3166,32 @@ def compute_point_cone_global_sizes(PETSc.DM dm):
...
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def mark_points_with_function_array(PETSc.DM plex,
+                                    PETSc.Section section,
+                                    PetscInt height,
+                                    np.ndarray[PetscInt, ndim=1, mode="c"] array,
+                                    PETSc.DMLabel dmlabel,
+                                    PetscInt label_value):
+
+    """Marks points in a DMLabel using indicator function array.
```
Comment (dham) [inline firedrake/cython/dmcommon.pyx:3180]: "using an indicator..."

File: firedrake/mesh.py [hunk 3]
```diff
@@ -2601,6 +2677,85 @@ def _on_rank_vertices(vertex_coords, parent_mesh):
...
+
+    if not isinstance(mesh, MeshGeometry):
+        raise TypeError(f"mesh must be a MeshGeometry, not a {type(mesh)}")
+    tmesh = mesh.topology
+    if isinstance(tmesh, VertexOnlyMeshTopology):
+        raise NotImplementedError("Currently does not work with VertexOnlyMesh")
+    elif isinstance(tmesh, ExtrudedMeshTopology):
+        raise NotImplementedError("Currently does not work with ExtrudedMesh; use RelabeledMesh() on the base mesh and then extrude")
+    if not isinstance(indicator_functions, (list, tuple)) or \
+       not isinstance(subdomain_ids, (list, tuple)):
```
Comment (dham) [inline firedrake/mesh.py:2704]: `isinstance(..., collections.abc.Sequence)`

Comment (dham) [reply firedrake/mesh.py:2704]: check for integer labels (`isinstance(..., numbers.Integral)`)

**Review by dham — CHANGES_REQUESTED** [review]

File: firedrake/cython/dmcommon.pyx (hunk 2 above)
Comment (ksagiyam) [reply firedrake/cython/dmcommon.pyx]: Fixed.

File: firedrake/mesh.py (hunk 3 above)
Comment (ksagiyam) [reply firedrake/mesh.py]: Fixed.

**Comment by nabw** [comment]
Hi, I would like to use this PR, but it looks like it can't be used for BCs. The following code:

`mesh = UnitSquareMesh(10, 10)
Vmarks = FunctionSpace(mesh, "HDiv Trace", 0)
x, y = SpatialCoordinate(mesh)
marks_fun = conditional(And(lt(x, 0.6), gt(x, 0.4)), 1.0, 0.0)
mesh = RelabeledMesh(mesh, [Function(Vmarks).interpolate(marks_fun)], [5])
assemble(1 * ds(5, domain=mesh)`

returns 0, and indeed if I add

`V = FunctionSpace(mesh, 'CG', 1)
bc = DirichletBC(V, Constant(1), 5)
f = Function(V)
bc.apply(f)`

the function instead interpolates in the entire domain the function marks_fun. Is this expected?

**Comment by ksagiyam** [comment]
Thanks for the feedback. The former is not expected, but the latter is.

For the latter, if we want to use this functionality to apply Dirichlet BCs (in the standard sense), the indicator function `marks_fun` should only mark boundary facets (with `y < .1` and `y > .9`, say).

For the former, the problem is that there are conflicting labels for the part you marked with `5` (e.g. on a part of the top boundary, you have the original label `4` and the new one `5`), and currently Firedrake does not allow for multiple labels (it just picks `4` and ignores `5`). This restriction is technically imposed in `mesh.py`:
```
class MeshTopology:
    ...
    def _facets(self, kind):
        ...
        markers = dmcommon.get_facet_markers(dm, facets)
        ...
```
A workaround is to remove the unwanted labels as the following:
```
from firedrake import *
import firedrake.cython.dmcommon as dmcommon


mesh = UnitSquareMesh(10, 10)
Vmarks = FunctionSpace(mesh, "HDiv Trace", 0)
x, y = SpatialCoordinate(mesh)
marks_fun = conditional(And(lt(x, 0.6), gt(x, 0.4)), 1.0, 0.0)
mesh = RelabeledMesh(mesh, [Function(Vmarks).interpolate(marks_fun)], [5])
mesh.topology_dm.clearLabelStratum(dmcommon.FACE_SETS_LABEL, 3)
mesh.topology_dm.clearLabelStratum(dmcommon.FACE_SETS_LABEL, 4)
print(assemble(1 * ds(5, domain=mesh)))
```
This prints `0.4` as expected.

I will think about what to do with this some time next week.

File: firedrake/mesh.py [hunk 4]
```diff
@@ -2857,6 +2944,89 @@ def _on_rank_vertices(vertex_coords, parent_mesh, tolerance):
     return np.compress(on_rank, vertex_coords, axis=0)
 
 
+def RelabeledMesh(mesh, indicator_functions, subdomain_ids, **kwargs):
+    """Construct a new mesh that has new subdomain ids.
+
+    :arg mesh: base :class:`~.MeshGeometry` object using which the
```
Comment (ReubenHill) [inline firedrake/mesh.py:2950]: I think this is what's causing doc build failure - I think this should be 
```
:class:`.MeshGeometry`
```

**Review by dham — APPROVED** [review]

**Review by dham — DISMISSED** [review]

**Comment by ksagiyam** [comment]
Fixed most of the link issues, but, for some, just used ``  ``.

**Review by connorjward — APPROVED** [review]
I think this will just need forcing through. Both tests and docs are failing but neither seems to be your fault.

**Comment by ksagiyam** [comment]
Thanks Connor. I will.

**Pull request merged**